File: 2023/python/solutions/day03.py
from solution import ProblemInput, Solution

Number = tuple[int, int, int, int]
Symbol = tuple[str, int, int]


class Solution03(Solution):
    def __init__(self) -> None:
        super().__init__(3)

    def _solve_version01(self, data: ProblemInput) -> int:
        numbers, symbols = parse(data)
        return sum(n[0] for n in numbers if is_part(n, symbols))

    def _solve_version02(self, data: ProblemInput) -> int:
        numbers, symbols = parse(data)
        return sum(gear_ratio(s, numbers) for s in symbols)


def parse(lines: list[str]) -> tuple[list[Number], list[Symbol]]:
    numbers = []
    symbols = []
    for i, line in enumerate(lines):
        number = ""
        number_start = -1
        for j, char in enumerate(line):
            if char in "0123456789":
                if number == "":
                    number_start = j
                number += char
            else:
                if number != "":
                    numbers.append((int(number), i, number_start, j - 1))
                    number = ""

                if char != "." and char != "\n":
                    symbols.append((char, i, j))

    return numbers, symbols


def is_part(number: Number, symbols: list[Symbol]) -> bool:
    _, num_row, left, right = number

    for sym, sym_row, sym_col in symbols:
        if sym_row == num_row and (sym_col == left - 1 or sym_col == right + 1):
            return True
        if (sym_row == num_row - 1 or sym_row == num_row + 1) and left - 1 <= sym_col <= right + 1:
            return True

    return False


def gear_ratio(symbol: Symbol, numbers: list[Number]) -> int:
    char, sym_row, sym_col = symbol

    if char != "*":
        return 0

    part_numbers = []
    for num, num_row, num_col_start, num_col_end in numbers:
        if sym_row == num_row and (sym_col == num_col_start - 1 or sym_col == num_col_end + 1):
            part_numbers.append(num)
            continue
        if (sym_row == num_row - 1 or sym_row == num_row + 1) and num_col_start - 1 <= sym_col <= num_col_end + 1:
            part_numbers.append(num)
            continue

    if len(part_numbers) != 2:
        return 0

    return part_numbers[0] * part_numbers[1]

File: 2023/rust/solutions/day03.rs
use crate::solution::{ProblemInput, ProblemResult, Solution};

#[derive(Debug)]
struct Number {
    value: u32,
    row: usize,
    col_start: usize,
    col_end: usize,
}

#[derive(Debug)]
struct Symbol {
    value: char,
    row: usize,
    col: usize,
}

pub struct Solution03 {}

impl Solution03 {
    fn parse(&self, lines: ProblemInput) -> (Vec<Number>, Vec<Symbol>) {
        let mut numbers = vec![];
        let mut symbols = vec![];
        for (i, line) in lines.iter().enumerate() {
            let mut num_str = String::from("");
            let mut number_start = 0;
            for (j, ch) in line.chars().enumerate() {
                if ch.is_ascii_digit() {
                    if num_str.is_empty() {
                        number_start = j;
                    }
                    num_str.push(ch);
                } else {
                    if !num_str.is_empty() {
                        numbers.push(Number {
                            value: num_str.parse().unwrap(),
                            row: i,
                            col_start: number_start,
                            col_end: j - 1,
                        });
                        num_str.clear();
                    }

                    if ch != '.' && ch != '\n' {
                        symbols.push(Symbol {
                            value: ch,
                            row: i,
                            col: j,
                        })
                    }
                }
            }

            if !num_str.is_empty() {
                numbers.push(Number {
                    value: num_str.parse().unwrap(),
                    row: i,
                    col_start: number_start,
                    col_end: line.len() - 1,
                });
            }
        }

        (numbers, symbols)
    }

    fn is_part(&self, num: &Number, symbols: &Vec<Symbol>) -> bool {
        for sym in symbols {
            if sym.row == num.row && (sym.col + 1 == num.col_start || sym.col == num.col_end + 1) {
                return true;
            }
            if (sym.row + 1 == num.row || sym.row == num.row + 1)
                && (num.col_start <= sym.col + 1 && sym.col <= num.col_end + 1)
            {
                return true;
            }
        }

        false
    }

    fn gear_ratio(&self, sym: &Symbol, numbers: &Vec<Number>) -> u32 {
        if sym.value != '*' {
            return 0;
        }

        let mut part_numbers = vec![];
        for num in numbers {
            if sym.row == num.row && (sym.col + 1 == num.col_start || sym.col == num.col_end + 1) {
                part_numbers.push(num.value);
                continue;
            }
            if (sym.row + 1 == num.row || sym.row == num.row + 1)
                && (num.col_start <= sym.col + 1 && sym.col <= num.col_end + 1)
            {
                part_numbers.push(num.value);
                continue;
            }
        }

        match part_numbers.len() {
            2 => {
                return part_numbers.iter().product();
            }
            _ => 0,
        }
    }
}

impl Solution for Solution03 {
    fn get_day(&self) -> u8 {
        3
    }

    fn solve_version01(&self, input: ProblemInput) -> ProblemResult {
        let (numbers, symbols) = self.parse(input);
        numbers
            .iter()
            .filter(|n| self.is_part(n, &symbols))
            .map(|n| n.value)
            .sum::<u32>()
            .try_into()
            .unwrap()
    }

    fn solve_version02(&self, input: ProblemInput) -> ProblemResult {
        let (numbers, symbols) = self.parse(input);
        symbols
            .iter()
            .map(|s| self.gear_ratio(s, &numbers))
            .sum::<u32>()
            .try_into()
            .unwrap()
    }
}
